clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
I think we both agree that buffering the fetched data will require much more thought and effort than throwing away data and re-fetch. Can you elaborate the benefit of buffering fetched data over throwing away data, that makes these effort worthwhile?

Specifically, could you explain why it is more natural to buffer the data? I think throwing away data is more intuitive -- user calls poll(); there is exception; user handles it; and try poll() again. It is common, for example in batch processing system, to start your job all over after exception and failure handling. Starting from where you left is really an optimization that comes later in time.

Here is another reason why buffering the fetched data can be problematic. Suppose the poll() caused exception,